File: devo_carre/src/main/java/com/test/devo_carre/domain/model/Reservation.java
package com.test.devo_carre.domain.model;

import java.time.Instant;
import java.util.UUID;

public record Reservation(
        UUID id,
        UUID eventId,
        UUID seatId,
        UUID userId,
        ReservationStatus status,
        Instant createdAt,
        Instant confirmedAt,
        Instant expiresAt,
        Long version
) {
    public Reservation confirm(Instant at) {
        if (status != ReservationStatus.PENDING) {
            throw new IllegalStateException("Reservation cannot be confirmed in status " + status);
        }
        return new Reservation(id, eventId, seatId, userId, ReservationStatus.CONFIRMED, createdAt, at, expiresAt, version);
    }

    public Reservation expire() {
        if (status != ReservationStatus.PENDING) {
            return this;
        }
        return new Reservation(id, eventId, seatId, userId, ReservationStatus.EXPIRED, createdAt, confirmedAt, expiresAt, version);
    }

    public Reservation cancel() {
        if (status != ReservationStatus.PENDING && status != ReservationStatus.CONFIRMED) {
            throw new IllegalStateException("Reservation cannot be cancelled in status " + status);
        }
        return new Reservation(id, eventId, seatId, userId, ReservationStatus.CANCELLED, createdAt, confirmedAt, expiresAt, version);
    }

    public boolean isActive(Instant now) {
        return status == ReservationStatus.CONFIRMED
                || (status == ReservationStatus.PENDING && expiresAt != null && expiresAt.isAfter(now));
    }
}

File: client-application/src/schemas/reservation.ts
import { z } from "zod";

const uuidSchema = z.string().uuid();

export enum ReservationStatus {
  PENDING = "PENDING",
  CONFIRMED = "CONFIRMED",
  CANCELLED = "CANCELLED",
  EXPIRED = "EXPIRED",
}

export const reservationBatchItemCommandSchema = z.object({
  eventId: uuidSchema,
  seats: z.array(uuidSchema).min(1, "At least one seat required"),
});

export const reservationViewSchema = z
  .object({
    id: uuidSchema,
    eventId: uuidSchema,
    seatId: uuidSchema,
    userId: uuidSchema,
    status: z.nativeEnum(ReservationStatus),
    createdAt: z.string(),
    confirmedAt: z.string().optional().nullable(),
    expiresAt: z.string().optional().nullable(),
    expires_at: z.string().optional().nullable(),
  })
  .transform((o) => ({
    id: o.id,
    eventId: o.eventId,
    seatId: o.seatId,
    userId: o.userId,
    status: o.status,
    createdAt: o.createdAt,
    confirmedAt: o.confirmedAt ?? null,
    expiresAt: o.expiresAt ?? o.expires_at ?? null,
  }));

export type ReservationBatchItemCommand = z.infer<
  typeof reservationBatchItemCommandSchema
>;
export type ReservationView = z.infer<typeof reservationViewSchema>;
